### ISSUE TITLE
Excellent work!

### ISSUE THREAD

**Issue by lirundong** [WangLeshanQAQ/learngit#1]
I do think this repository is an elegant and faithful reimplementation of the DoReFa paper. Belowing are some minor concerns to me:

1. Please try to list some numeracal results and compare them with the origin paper, e.g. the accuracy of AlexNet on SVHN dataset (Table 1 in paper). This is meant to ensure the correctness of your implementation. You can utilize the [torchvision SVHN dataset](https://pytorch.org/docs/stable/torchvision/datasets.html#svhn). I suggest that firstly train a FP32 baseline and compare it with the last row of Table 1, then go for W1/A2/G32 setting, finally go for W1/A2/G4 setting;
2. There is a minor mistake in [gradient quantizer implementation](https://github.com/WangLeshanQAQ/learngit/blob/63f361cf6c8a0b9d051b9dda13cf6c0272c124e3/utils/quantizer.py#L72). Please note the equation 12 in origin paper, the absolute magnitude is taken from **each instance within the minibatch**:
   > and the maximum is taken over all axis of the gradient tensor dr except for the mini-batch axis (therefore each instance in a mini-batch will have its own scaling factor)
   
   so you can pass `dim=0` option to [`torch.max`](https://pytorch.org/docs/stable/torch.html#torch.max). Be aware of the return values when `dim` is assigned;
3. Please try to use `.gitignore` to filter out the `.DS_Store` and `__pycache__` files. They are not meant to tracked by git.

After all, very very excellent work. Once you done the numeracal camparison, we can have some conversations and go for broader directions.

**Comment by lirundong** [comment]
Updated review:

I compared this reimplementation with the [official implementation](https://github.com/tensorpack/tensorpack/blob/master/examples/DoReFa-Net/svhn-digit-dorefa.py), there are some differences:

1. `DigitNet_Q` is not consistent with the origin paper: the building blocks should be in `Conv -> BN -> [MaxPool] -> ReLU (clip 0, 1) -> QuantAct` order as [here](https://github.com/tensorpack/tensorpack/blob/a7f4094d86468a0ff0dae51f0640f7f373d231f2/examples/DoReFa-Net/svhn-digit-dorefa.py#L84-L88), rather than `Conv -> BN -> ReLU -> MaxPool -> Clip 0, 1 -> QuantAct` in [your implementation](https://github.com/WangLeshanQAQ/learngit/blob/91edc1b232468cb3bea95ce55025037067556f2d/models/svhn_digit_dorefa.py#L31-L37);
2. The first Conv layer don't have following BN, as [here](https://github.com/tensorpack/tensorpack/blob/a7f4094d86468a0ff0dae51f0640f7f373d231f2/examples/DoReFa-Net/svhn-digit-dorefa.py#L76);
3. [Your configuration](https://github.com/WangLeshanQAQ/learngit/blob/91edc1b232468cb3bea95ce55025037067556f2d/main_svhn.py#L96) of Adam optimizer is not consistent with [the official one](https://github.com/tensorpack/tensorpack/blob/a7f4094d86468a0ff0dae51f0640f7f373d231f2/examples/DoReFa-Net/svhn-digit-dorefa.py#L132). You should not apply weight decay. Please refer to [this note](https://cs231n.github.io/neural-networks-3/) for more details;
4. You can implement exponential LR decay by [ExponentialLR](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.ExponentialLR). Please refer to TensorFlow [`exponential_decay` doc](https://www.tensorflow.org/api_docs/python/tf/train/exponential_decay) for the underlaying scheme. 

Please try to fix these issues and try again.
